### ISSUE TITLE
Loader emits unnecessary dependencies to compilation

### ISSUE THREAD

**Issue by colinskow** [s-panferov/awesome-typescript-loader#351]
**Expected behavior:** the loader should only emit files actually in the module (or reference) dependency tree as file dependencies to Webpack.

**Actual behavior:** the loader emits every single TypeScript file in the project as a dependency regardless of the entries. This can cause performance problems in watch mode.

I have one entry which is a TypeScript file with no dependencies, and at-loader is emitting hundreds of files for Webpack to watch. In my use case I am building an electron app with separate builds for the main and renderer processes. When a dependency of main changes I want to restart the main process, and when a dependency of the renderer changes I want to refresh the browser. Right now this impossible because at-loader causes Webpack to watch every single .ts file in the project, including `node_modules`.

I tracked the offending code down to [`src/instance.ts#L368`](https://github.com/s-panferov/awesome-typescript-loader/blob/58a161c0d25b3353c453e9ec4f322b6874c61866/src/instance.ts#L368):

```js
Array.prototype.push.apply(compilation.fileDependencies, files.map(path.normalize));
```

When I comment out this line the loader behaves as expected and only files from the actual dependency tree of the entries are added to the Webpack compilation as dependencies.

The correctly needed dependencies are added [here](https://github.com/s-panferov/awesome-typescript-loader/blob/0f6f1276939d958b23142af61e6e6e81ad22024e/src/index.ts#L87), and the code above in `instance.ts` is wasteful and unnecessary.

**Comment by colinskow** [comment]
OK, I realized there is a fatal flaw in this approach that lazy loaded modules (such as from Angular 2) don't get emitted as dependencies in this scenario.

I think a better solution is to probably specify the exact entry files that need to be parsed under "files" in `tsconfig.json`. This way the TypeScript compiler doesn't parse anything unnecessary.

**Comment by colinskow** [comment]
What I need to resolve this issue is a way to pass in a list of filenames, which will be merged into the `tsconfig.json` as `files`.

Furthermore, it would be nice to be able to pass the loader an object which will be merged on top of the contents of `tsconfig.json`. I saw there is a `configFileContent` options available in `instance.ts` but it didn't work as expected.

**Comment by s-panferov** [comment]
Closed in favour of #364
